src/games/strategy/common/swing/SwingComponents.java

```diff
@@ -91,17 +100,31 @@ public static void promptUser(final String title, final String message, final Ru
 
   public static void promptUser(final String title, final String message, final Runnable confirmedAction,
       final Runnable cancelAction) {
-    SwingUtilities.invokeLater(() -> {
-      int response = JOptionPane.showConfirmDialog(null, message, title,
-          JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
-      boolean result = response == JOptionPane.YES_OPTION;
-
-      if (result) {
-        confirmedAction.run();
-      } else {
-        cancelAction.run();
+
+    boolean showMessage = false;
+    synchronized (visiblePrompts) {
```

Comment:
Are you sure this would prevent that? Can you explain the scenario?

Reading through it, I'm not sure that it handles all cases and might to need expand the synchronized scope. Couldn't you have one thread run through the sync block then another flip the showMessage flag to false which would stomp on showing the first prompt?

Potentially even worse is that message wouldn't be removed in the second block and then its already in the list so couldn't be re-added in the sync block.